File: node/src/sensor/cold.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a cold sensor.
 *
 * @internal
 */
export enum ColdProperty {
  /** Whether abnormal cold is detected. */
  Detected = 'detected',
}

/**
 * Property values of a cold sensor.
 *
 * @internal
 */
export interface ColdSensorProperties {
  [ColdProperty.Detected]: boolean;
}

/** Read-only proxy interface for a cold sensor. */
export interface ColdSensorLike extends SensorLike {
  readonly type: SensorType.Cold;
  readonly onPropertyChanged: Observable<PropertyChangeOf<ColdSensorProperties>>;

  getValue(property: ColdProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Cold alarm sensor. */
export class ColdSensor<TStorage extends object = Record<string, any>> extends Sensor<ColdSensorProperties, TStorage> {
  readonly type = SensorType.Cold;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Cold Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [ColdProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report abnormal cold detection state.
   *
   * @param value - True when cold is currently detected.
   *
   * @example
   * ```ts
   * cold.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [ColdProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link ColdSensor}. */
export const coldMeta = defineSensor({
  type: SensorType.Cold,
  category: SensorCategory.Sensor,
  assignmentKey: 'cold',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [ColdProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: ColdProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [ColdProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: ColdProperty.Detected,
    commandProperty: ColdProperty.Detected,
    deviceClass: 'cold',
  },
});

File: python/camera_ui_sdk/sensor/cold.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class ColdProperty(StrEnum):
    """Property names of a cold sensor."""

    Detected = "detected"
    """Whether abnormal cold is detected."""


class ColdSensorProperties(TypedDict):
    """Property values of a cold sensor."""

    detected: bool


class ColdPropertyChangeData(TypedDict):
    """Property change payload emitted on ColdSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a ColdProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class ColdSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a cold sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Cold

    @property
    def onPropertyChanged(self) -> Observable[ColdPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[ColdProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class ColdSensor(Sensor[ColdSensorProperties, TStorage, str], Generic[TStorage]):
    """Cold alarm sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Cold Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({ColdProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Cold

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report abnormal cold detection state.

        Args:
            value: True when cold is currently detected.

        Example:
            ```python
            cold.setDetected(True)
            ```
        """
        self._write_state({ColdProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""
